### PR TITLE
feat: added conditions to update / delete + pagination to list

### PR DESCRIPTION
**Pull request by MacMcIrish**

### CODE REVIEW

File: src/modules/model.js [hunk 1]
```diff
@@ -147,21 +155,55 @@ class Model {
...
+    const splitFields = objectFields.split(fields);
+    const fieldsContainsId = splitFields.includes('id');
+    if (!fieldsContainsId) {
+      splitFields.push('id');
+    }
+
+    const {
+      lastEvaluatedKey,
+      scanIndexForward = true,
+      limit: queryLimit = limit,
```
Comment (MacMcIrish) [inline src/modules/model.js]: Awkward, in that we haven't done param renaming with a default before. I don't necessarily think it's bad, feel free to disagree though :)

File: src/util/paging.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+const objectDecode = base64 => JSON.parse(Buffer.from(base64, 'base64').toString('utf8'));
+
+module.exports.fromCursor = (cursor) => {
+  const {
+    lastEvaluatedKey, scanIndexForward, currentPage, limit
+  } = cursor === null ? {} : objectDecode(cursor);
+  return {
+    lastEvaluatedKey, scanIndexForward, currentPage, limit
+  };
+};
```
Comment (simlu) [inline src/util/paging.js]: what happens if the cursor is invalid?

File: src/modules/model.js [hunk 3]
```diff
@@ -147,21 +155,55 @@ class Model {
...
     this._before();
+    const splitFields = objectFields.split(fields);
+    const fieldsContainsId = splitFields.includes('id');
+    if (!fieldsContainsId) {
+      splitFields.push('id');
+    }
+
+    const {
+      lastEvaluatedKey,
+      scanIndexForward = true,
```
Comment (simlu) [inline src/modules/model.js]: is this still necessary?

File: src/modules/model.js [hunk 4]
```diff
@@ -147,21 +155,55 @@ class Model {
...
+      limit: queryLimit = limit,
+      currentPage = null
+    } = fromCursor(cursor);
     const iterator = this.mapper.query(this.MapperClass, indexMap, {
       indexName,
-      projection: objectFields.split(fields)
+      projection: splitFields,
+      startKey: lastEvaluatedKey,
+      scanIndexForward,
```
Comment (simlu) [inline src/modules/model.js]: probably just inline?

File: src/modules/model.js [hunk 5]
```diff
@@ -147,21 +155,55 @@ class Model {
...
     // eslint-disable-next-line no-underscore-dangle
     this._before();
+    const splitFields = objectFields.split(fields);
+    const fieldsContainsId = splitFields.includes('id');
+    if (!fieldsContainsId) {
+      splitFields.push('id');
+    }
+
+    const {
+      lastEvaluatedKey,
```
Comment (simlu) [inline src/modules/model.js]: make this explicit as null and set correctly below using assign? Not super important though

File: src/util/paging.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+const assert = require('assert');
+
+const objectEncode = obj => Buffer.from(JSON.stringify(obj)).toString('base64');
+
```
Comment (simlu) [inline src/util/paging.js]:
```suggestion
```

File: src/util/paging.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+const objectDecode = base64 => JSON.parse(Buffer.from(base64, 'base64').toString('utf8'));
+
+module.exports.fromCursor = (cursor) => {
+  let cursorPayload = {};
+  if (cursor !== null) {
+    try {
+      cursorPayload = objectDecode(cursor);
+    } catch (err) {
+      assert(err.name === 'SyntaxError' && err.message.startsWith('Unexpected token'), err);
```
Comment (simlu) [inline src/util/paging.js]: Is there is a problem with the returned/transmitted token the client would page indefinite...

I would just throw a new error "InvalidPageCursor" or something like that and then handle it outside the library

Comment (MacMcIrish) [reply src/util/paging.js]: Oh damn, that's a good point. I'll do exactly that

File: src/modules/model.js [hunk 8]
```diff
@@ -66,7 +68,7 @@ class Model {
     let resp;
     try {
       resp = await this.mapper.get(new this.MapperClass({ id }), {
-        projection: objectFields.split(fields),
+        projection: projectionUtil.format(fields),
```
Comment (simlu) [inline src/modules/model.js]: Honestly think this should only handle arrays

File: src/modules/model.js [hunk 9]
```diff
@@ -147,21 +155,54 @@ class Model {
...
+      limit: queryLimit = limit,
+      currentPage = null
+    } = fromCursor(cursor);
+    const iterator = this.mapper.query(this.MapperClass, indexMap, Object.assign({
       indexName,
-      projection: objectFields.split(fields)
-    });
-    const resp = [];
+      projection,
+      scanIndexForward,
```
Comment (simlu) [inline src/modules/model.js]: I removed the comment I left this morning, I think it's a valid parameter for the list function

**Comment by krackerman** [comment]
@dependabot rebase

**Pull request merged**